.github/workflows/container.yml

```diff
@@ -195,24 +195,57 @@ jobs:
         run: podman images | grep 'ghcr.io/parca-dev/parca-agent'
 
       - name: Install cosign
-        if: ${{ github.event_name != 'pull_request' }}
         uses: sigstore/cosign-installer@48866aa521d8bf870604709cd43ec2f602d03ff2 # tag=v2.4.1
 
       - name: Login to registry
-        if: ${{ github.event_name != 'pull_request' }}
         run: |
           echo "${{ secrets.GITHUB_TOKEN }}" | podman login -u parca-dev --password-stdin ghcr.io
           echo "${{ secrets.QUAY_PASSWORD }}" | cosign login -u "${{ secrets.QUAY_USERNAME }}" --password-stdin quay.io
 
       - name: Install crane
-        if: ${{ github.event_name != 'pull_request' }}
         uses: imjasonh/setup-crane@5146f708a817ea23476677995bf2133943b9be0b # tag=v0.1
 
-      - name: Push and sign container
-        if: ${{ github.event_name != 'pull_request' }}
+      - name: Push container for end-to-end test
+        run: make push-container-test
+
+      - name: Push and sign container(when merged)
         env:
           COSIGN_EXPERIMENTAL: true
         run: |
-          make push-container
-          make sign-container
-          make push-signed-quay-container
+          if [[ ${{ github.event_name }} != 'pull_request' ]]; then
+              make push-container
+              make sign-container
+              make push-signed-quay-container
+          fi
+
+  run-end-to-end-tests:
+    name: run end-to-end tests
+    # e2e tests need nested virtualisation which is only available
+    # on macos runners when using GitHub-hosted runners.
+    runs-on: macos-12
+    needs: build-and-push-container
+    permissions:
+      contents: read
+    steps:
+      - name: Check out the code
+        uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b # tag=v3.0.2
+
+      - name: Set up Go
+        uses: actions/setup-go@84cbf8094393cdc5fe1fe1671ff2647332956b1a # tag=v3.2.1
+        with:
+          go-version-file: .go-version
+          cache: true
```

Comment:
We should not say `cache: true` these packages are different from our usual packages. We've seen that this could cause unnecessary errors.